lib/esm-utils.js

```diff
@@ -0,0 +1,27 @@
+// This file is allowed to use async/await because it is not exposed to browsers (see the `eslintrc`),
+// and Node supports async/await in all its non-dead version.
+
+const url = require('url');
+const path = require('path');
+
+exports.requireOrImport = file => {
+  try {
+    file = path.resolve(file);
+    return require(file);
+  } catch (err) {
```

Comment:
Is there no other way to find out, extension `*mjs` or `type: module`.
Or some internal Node property?

Reply:
yes, we should be able to short-circuit this a bit:

1. Use `require` with `.cjs` files
1. Use `import()` with `.mjs` files

Other than that, I'm not seeing that the package scope resolution method is exposed anywhere (I'll need to look around in the node source).  So if we wanted to consume `type` in `package.json`, we'd need to use [read-pkg-up](https://github.com/sindresorhus/read-pkg-up) and manage our own cache, which seems gross.

Also, the cache-clearing method that `Mocha` exposes won't work with ESM...

Reply:
OK, `import()` will handle _both_ ESM and CJS, so we don't need to try/catch _at all_ here.

Reply:
Cache-clearing using ESM is an unanswered question atm, and it's best to not worry our pretty heads about it.

Reply:
this means we can't support "watch" mode until that's solved in Node core.  there is no cache available to clear.  right now, using ESM for tests has two major drawbacks due to this missing feature:

1. You can't hot-reload a module for "watch" mode
2. You can't hot-reload a module to do module-level mocking (like with [rewiremock](https://npm.im/rewiremock))

Reply:
@juergba: no, there is currently no way to determine if a file is ESM or CJS, although Node's Module Working Group is deliberating on that (https://github.com/nodejs/node/issues/30514)

@boneskull, while it is true that `import()` should handle CJS, remember that _all_ the importing/requiring in Node will go through this function. I was hesitant in making `import()` the way we import all test files in Mocha, as who knows what hidden bugs and weird behavior lurks there.

So I opted for this weird `require`/`import` style to solve both problems: not knowing what type the file is, and fear of unknown behavior when `import`-ing CJS. I think it's a good way to start this, given that practically 100% of test files out there are CJS. We can revisit this decision in a couple of months once people really start using this feature. Till then, I believe using `require` optimistically and falling back on `import` is the better decision.

Reply:
@boneskull - yup, watch mode cannot yet be implemented for ESM, unless we use a cache-busting mechanism (ESM paths are URLs, so we can add arbirtrary query parameters to them) to it.

I have pinged the modules working group on this problem (https://github.com/nodejs/modules/issues/307#issuecomment-558462605).

Reply:
Ok, I could live with this (ugly) solution to start with, since Node's v13.2.0 docu still labels the esm module with "Stability: 1 - Experimental". But then please add  a short comment in order to remind us there is still some work to do.

Reply:
Thinking about it I vote for:
- short-circuit: require() for .cjs / import() for .mjs
- then your "weird require/import style" since we still have experimental status.

Reply:
> So I opted for this weird require/import style to solve both problems: not knowing what type the file is, and fear of unknown behavior when import-ing CJS. I think it's a good way to start this, given that practically 100% of test files out there are CJS. We can revisit this decision in a couple of months once people really start using this feature. Till then, I believe using require optimistically and falling back on import is the better decision.

I think this is fair until we've seen `import()` in wider use.